### ISSUE TITLE
Kivy on pyodide?

### ISSUE THREAD

**Issue by TensorTom** [iodide-project/pyodide#1238]
This might sound crazy but if wanting Kivy in pyodide is crazy, I don't want to be sane. Maybe it would be a big task. Kivy uses Cython, so that gives me hope. It also has GPU code but idk how they approached it. Maybe it doesn't matter since it's running on Cython? I want to find out.

**Comment by ehfd** [comment]
Making it to a Pyodide package is one thing. The other is visualizing the GUI on the web. I saw such attempts in Java so it might be possible. Kivy development teams think they are possible but are negative about such attempt and they said to use a real web framework. It would work but not "modern" web at all.

**Comment by TensorTom** [comment]
> Making it to a Pyodide package is one thing. The other is visualizing the GUI on the web. I saw such attempts in Java so it might be possible. Kivy development teams think they are possible but are negative about such attempt and they said to use a real web framework. It would work but not "modern" web at all. If you really need it, you need to it yourself (or fund someone).

I wonder why they are negative about it. We have webgl.

**Comment by TensorTom** [comment]
I found a thread at the Kivy repo which starts with [the message](https://github.com/kivy/kivy/issues/4469#issuecomment-521809981):

> Here's a experimental implementation of kivy running in Pyodide. You can explore with it here: https://alpha.iodide.io/notebooks/2882/.
> 
> A few things to keep in mind:
> 
> Pyodide is still highly experimental and so is my implementation of kivy it.
> Read the explanation text in the notebook.
> The kivy window is shown at the bottom of the text to the right, so scroll all the way down after you run the app.
> Make sure to wait a little before running each cell, especially the first time before you run the actual app. Otherwise you may get OSError: broken data stream when reading image file. If you do get that error, just refresh the page and try again, or run the code again.
> There seems to be a transparency issue where the background is not cleared.
> Text looks awful.
> This is run using native js api (I wrote a new window and gl provider using the js native API), while this prevents any compilation issues it probably (maybe) is slower than the sdl2.
> I did not implement the keyboard and clipboard yet. Keyboard should be easy, clipboard we'll have to work out the permissions.

**Comment by matham** [comment]
I don't think we are negative about it (in fact you quoted from my implementation post), it's just that running kivy in the browser, beyond for experimental reasons has downsides compared to any native framework.

It's one thing if you're doing e.g. data science where most of the code is python and you want to visualize it in the browser. But running a actual app in the browser using Kivy just seems counterintuitive, especially at the present while pyodide is in alpha. The downloads/load time will be large, the interface will not look at all like the web, and there are many components that if you use it in your app you will need to adapt it (threading, async, clipboard just to name a few, and if you need audio/video well good luck). 

So to get it to where running it in the browser is good experience will take a fair amount of work, which currently no one seems to have the time for. So even if we do have a pyodide implementation, you shouldn't expect to just be able to make your app work in it without significant changes. And at that point it seems simpler to figure out how to distribute app binaries.

But of course it would be nice to get kivy working in it. However, there are technical blockages (e.g. https://github.com/iodide-project/pyodide/issues/677 was the most recent issue), or at least last time I checked.

**Comment by ehfd** [comment]
@matham Yes I wrote my initial answer after seeing [this](https://stackoverflow.com/questions/42183014/how-do-i-deploy-a-kivy-gui-application-as-a-webapp-in-a-web-browser). Your notebook looks like a good start.

**Comment by TensorTom** [comment]
> I don't think we are negative about it (in fact you quoted from my implementation post), it's just that running kivy in the browser, beyond for experimental reasons has downsides compared to any native framework.
> 
> It's one thing if you're doing e.g. data science where most of the code is python and you want to visualize it in the browser. But running a actual app in the browser using Kivy just seems counterintuitive, especially at the present while pyodide is in alpha. The downloads/load time will be large, the interface will not look at all like the web, and there are many components that if you use it in your app you will need to adapt it (threading, async, clipboard just to name a few, and if you need audio/video well good luck).
> 
> So to get it to where running it in the browser is good experience will take a fair amount of work, which currently no one seems to have the time for. So even if we do have a pyodide implementation, you shouldn't expect to just be able to make your app work in it without significant changes. And at that point it seems simpler to figure out how to distribute app binaries.
> 
> But of course it would be nice to get kivy working in it. However, there are technical blockages (e.g. #677 was the most recent issue), or at least last time I checked.

Thanks for the reply. I understand the hurtles. I would point out though that similar things have been done before. Just not after the fact, having to tack on browser support years into a project. So point taken in that respect. Still would be fun though.

**Comment by tytgatlieven** [comment]
@TensorTom , @matham 

My use case is:
* I want an app to be able to run without the need to install it on the device. The need to install requires one to go through one of the app stores, which are monopolized.
* I don't want to have to write multiple apps on multiple platforms
* Kivy is technologically sound, and has a strong community. In my opinion this might become The app platform if a webbrowser is one of the supported platforms

I do realize that this is not a simple task. However, either we look at the possible benefits and figure out ways to do the needed work, or we only look at the needed work and do not try. I'm in favour of the first. There are no guarantees that we will succeed, but at the minimum we tried and failed doing great things. But what if we pull it off.

Last academic year I had 2 students working on figuring out which would be the good way to get kivy to work. We ended op trying to build the work @matham did on current pyodide. We succeeded in building everything and fixed the first few errors when running in the browser :-), however, no running Kivy yet :-(
This year I have 2 students working on this for a course, which will also do their internship in my company on this topic. At the moment they are working on dockerizing the complete setup in order to make collaboration on this work significantly easier. We will certainly share this on github once we have this part rounded up

We have only limited knownledge about the internals of kivy. Hence we can certainly use all the help we can get.
Anybody willing to help?

**Comment by FilipeMarch** [comment]
I would like to know if [PyScript](https://github.com/pyscript/pyscript) updates the current board in any way, i.e., if there is any hope for running Kivy apps on Browser due to do the existence of PyScript.

**Comment by hoodmane** [comment]
Well the dynamic linking bugs that @matham mentioned are most likely fixed now and if they aren't fixed I have very good techniques for that type of dynamic linker error after fixing many many many such errors in scipy (on the order of 50 to 100).

So I suspect the technical build blockers are resolvable but whether or not there's sufficient community interest to do the UI work needed is another question.

**Comment by hoodmane** [comment]
Oh so you don't have a reproduction on the error?
